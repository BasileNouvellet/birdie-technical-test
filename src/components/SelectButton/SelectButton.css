.select-button {
    width: 30%;
    margin: 3em auto;
}
