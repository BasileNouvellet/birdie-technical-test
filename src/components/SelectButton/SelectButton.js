// @flow

import * as React from 'react';
import { connect } from 'react-redux';
import Select from 'react-select';

import ErrorMessage from '../ErrorMessage';

import { TableActions } from '../Table';

import './SelectButton.css';

import type {
  TableStateColumnsType,
  TableStateType,
  TableStateVariableType,
  TableStateErrorType,
} from '../Table';

type MappedStatePropsType = {|
  columns: TableStateColumnsType,
  variable: TableStateVariableType,
  error: TableStateErrorType,
|};
type MappedDispatchPropsType = {|
  fetchColumns: () => void,
  fetchData: (variable: TableStateVariableType) => void,
  setVariable: (variable: TableStateVariableType) => void,
|};
type OwnPropsType = {||};
type PropsType = MappedStatePropsType & MappedDispatchPropsType & OwnPropsType;

// overwrite some of the default styles of React Select
const selectStyles = {
  option: provided => ({
    ...provided,
    textAlign: 'start',
  }),
};

// export unconnected component for test purposes with Jest
export class SelectButtonUnconnected extends React.Component<PropsType> {
  componentDidMount() {
    const { fetchColumns } = this.props;
    fetchColumns(); // fetch columns names in did mount
  }

  getOptions() {
    const { columns } = this.props;

    return columns
      // .filter(column => column !== 'age')
      .concat('THIS WILL FAIL') // add non existing column name to test error handling
      .sort() // sort alphabetically
      .map(column => ({ // map to React Select options structure
        value: column,
        label: column.toLocaleString(), // how it is going to be displayed to the user
      }));
  }

  handleChange = (newOption) => {
    const {
      variable, setVariable, fetchData,
    } = this.props;

    const { value: newValue } = newOption;

    if (variable !== newValue) {
      setVariable(newValue); // set new variable value in the redux store
      fetchData(newValue); // start fetching the data corresponding to the new variable
    }
  };

  // ------------------------------------------- Render -------------------------------------------
  render(): React.Element<*> {
    const { error } = this.props;
    if (error.columns) return <ErrorMessage />;

    const options = this.getOptions();

    return (
      <Select
        className="select-button"
        styles={selectStyles}
        placeholder="Select a variable..."
        onChange={this.handleChange}
        options={options}
      />
    );
  }
}

const mapStateToProps = (state: TableStateType): MappedStatePropsType => ({
  columns: state.columns,
  variable: state.variable,
  error: state.error,
});

const mapDispatchToProps = (dispatch: *): MappedDispatchPropsType => ({
  fetchColumns: () => {
    dispatch(TableActions.fetchColumnsAction());
  },
  fetchData: (variable: TableStateVariableType): void => {
    dispatch(TableActions.fetchDataAction(variable));
  },
  setVariable: (variable: TableStateVariableType): void => {
    dispatch(TableActions.setVariableAction(variable));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SelectButtonUnconnected);
