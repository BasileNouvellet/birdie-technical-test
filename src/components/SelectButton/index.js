// @flow

import SelectButton from './SelectButton';

export default SelectButton;
