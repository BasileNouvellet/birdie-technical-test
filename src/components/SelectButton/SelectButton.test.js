// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

import { SelectButtonUnconnected as SelectButton } from './SelectButton';

describe('SelectButton component', () => {
  let wrapper;

  const reduxStoreProps = {
    // map state to props
    columns: ['column1', 'column2', 'column3'],
    variable: 'column2',
    error: {
      column: false,
      data: false,
    },
    // map dispatch to props
    fetchColumns: jest.fn(),
    fetchData: jest.fn(),
    setVariable: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(
      <SelectButton
        columns={reduxStoreProps.columns}
        variable={reduxStoreProps.variable}
        error={reduxStoreProps.error}
        fetchColumns={reduxStoreProps.fetchColumns}
        fetchData={reduxStoreProps.fetchData}
        setVariable={reduxStoreProps.setVariable}
      />,
    );
  });

  it('should shallow render without crashing', () => {}); // done in beforeEach()

  it('should match snapshot', () => { expect(wrapper).toMatchSnapshot(); });
});
