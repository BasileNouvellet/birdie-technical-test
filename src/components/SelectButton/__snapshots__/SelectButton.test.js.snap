// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectButton component should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectButtonUnconnected
    columns={
      Array [
        "column1",
        "column2",
        "column3",
      ]
    }
    error={
      Object {
        "column": false,
        "data": false,
      }
    }
    fetchColumns={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    fetchData={[MockFunction]}
    setVariable={[MockFunction]}
    variable="column2"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "className": "select-button",
      "defaultInputValue": "",
      "defaultMenuIsOpen": false,
      "defaultValue": null,
      "onChange": [Function],
      "options": Array [
        Object {
          "label": "THIS WILL FAIL",
          "value": "THIS WILL FAIL",
        },
        Object {
          "label": "column1",
          "value": "column1",
        },
        Object {
          "label": "column2",
          "value": "column2",
        },
        Object {
          "label": "column3",
          "value": "column3",
        },
      ],
      "placeholder": "Select a variable...",
      "styles": Object {
        "option": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "className": "select-button",
        "defaultInputValue": "",
        "defaultMenuIsOpen": false,
        "defaultValue": null,
        "onChange": [Function],
        "options": Array [
          Object {
            "label": "THIS WILL FAIL",
            "value": "THIS WILL FAIL",
          },
          Object {
            "label": "column1",
            "value": "column1",
          },
          Object {
            "label": "column2",
            "value": "column2",
          },
          Object {
            "label": "column3",
            "value": "column3",
          },
        ],
        "placeholder": "Select a variable...",
        "styles": Object {
          "option": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
